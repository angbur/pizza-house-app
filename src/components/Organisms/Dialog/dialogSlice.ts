import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from 'store/store';

export enum FormType {
  register = 'register',
  login = 'login',
  orderDetail = 'orderDetail',
}

type DialogState = {
  isOpen: boolean;
  formType: FormType | undefined;
};

const initialState: DialogState = {
  isOpen: false,
  formType: undefined,
};

const dialogSlice = createSlice({
  name: 'dialog',
  initialState,
  reducers: {
    openDialog: (state, action: PayloadAction<FormType>) => {
      state.isOpen = true;
      state.formType = action.payload;
    },
    closeDialog: (state) => {
      state.isOpen = false;
      state.formType = undefined;
    },
  },
});

export const { openDialog, closeDialog } = dialogSlice.actions;

export default dialogSlice.reducer;

export const selectDialogState = (state: RootState) => state.dialog;
