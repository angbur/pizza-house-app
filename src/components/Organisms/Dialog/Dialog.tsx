import Button from 'components/Atoms/Button/Button';
import Typography from 'components/Atoms/Typography/Typography';
import { ThemeContext } from 'components/Theme/ThemeContext';
import { useAppDispatch, useAppSelector } from 'hooks';
import { useContext, useState } from 'react';
import styled from 'styled-components';
import { getDialogElement } from './dialog.utils';
import { closeDialog, FormType, openDialog, selectDialogState } from './dialogSlice';
import CloseIcon from 'assets/icon/cancel-cross.svg';
import { useRegisterMutation } from 'services/user';

const Modal = styled.div`
  position: fixed;
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  z-index: 100;
`;

const Overlay = styled.div`
  position: absolute;
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  z-index: 101;
  background: rgba(0, 0, 0, 0.85);
`;

type ModalProps = {
  isOpen: boolean;
};

const ModalContent = styled.div<ModalProps>`
  z-index: 102;
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  max-height: 90%;
  overflow: auto;
  background: ${(props) => props.theme?.palette.light};
  box-shadow: 0 1px 5px rgba(0, 0, 0, 0.7);
  border-radius: 4px;
  width: 400px;
  display: ${(props) => (props.isOpen ? 'block' : 'none')};
  & > .dialog-header {
    display: flex;
    justify-content: center;
    border-bottom: 1px solid ${(props) => props.theme?.palette.primary};
    padding: 1.625rem;
    & > button {
      position: absolute;
      top: 16px;
      right: 16px;
    }
  }
`;

const ModalActions = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  gap: 1rem;
  margin-bottom: 2rem;
`;

const Dialog = () => {
  const dispatch = useAppDispatch();
  const theme = useContext(ThemeContext);
  const { isOpen, formType } = useAppSelector(selectDialogState);
  const [data, setData] = useState({
    firstName: '',
    lastName: '',
    login: '',
    email: '',
    password: ''});
    const [register, { isLoading }] = useRegisterMutation();

  if (!formType) return null;
  const dialogDetail = getDialogElement(formType);

  const handleClose = () => {
    dispatch(closeDialog());
  };

  const handleRegister = () => {
    dispatch(closeDialog);
    dispatch(openDialog(FormType.register));
  };

  const handleLogin = () => {
    dispatch(closeDialog);
    dispatch(openDialog(FormType.login));
  };

  const handleSubmitRegister = () => {
   register(data)
  };

  const handleSubmitLogin = () => {
    null
  };

  const handleChange = ({
    target: { name, value },
  }: React.ChangeEvent<HTMLInputElement>) => {
    setData(
      {
        ...data,
        [name.trim()]: value
      })
  };

  return (
    <Modal>
      <Overlay onClick={handleClose} />
      <ModalContent theme={theme} isOpen={isOpen}>
        {dialogDetail && (
          <>
            <div className={'dialog-header'}>
              <Button
                variant={'icon-button'}
                className={'header_actions_button--mobile'}
                onClick={handleClose}
                style={{ ...{ padding: 0 } }}
              >
                <img src={CloseIcon} height={'12px'} alt={''} />
              </Button>
              <Typography
                variant='dialog-title'
                color='secondary'
                style={{ ...{ textTransform: 'uppercase' } }}
              >
                {dialogDetail.title}
              </Typography>
            </div>
            <div>{dialogDetail.component && dialogDetail.component({handleChange: handleChange})}</div>
            <ModalActions>
          
            { formType === 'register' ? (
              <>
                <Button variant='primary-light' size='md' type='submit' onClick={handleSubmitRegister}>
                  Create Account
                </Button>
                  <Typography variant='paragraph' color='secondary'>
                  If you already have an account just
                  <Button variant='button-text-light' size='sm' onClick={handleRegister}>
                    SIGN IN
                  </Button>
                </Typography>
              </>
            ) : formType === 'login' ? (
              <>
                 <Button variant='primary-light' size='md' type='submit' onClick={handleSubmitLogin}>
                  Log In
                </Button>
                  <Typography variant='paragraph' color='secondary'>
                    If you do not have an account 
                  <Button variant='button-text-light' size='sm' onClick={formType === 'register' ? handleRegister : handleLogin}>
                    SIGN UP
                  </Button>
                </Typography>
              </>
            ) : null}
          </ModalActions>
          </>
        )}
      </ModalContent>
    </Modal>
  );
};

export default Dialog;