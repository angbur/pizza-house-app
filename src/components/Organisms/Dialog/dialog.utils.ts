import { ChangeEventHandler, ReactNode } from 'react';
import Login from './components/Login';
import LoginSuccess from './components/LoginSuccess';
import OrderDetail from './components/OrderDetail';
import Register from './components/Register';
import RegisterSuccess from './components/RegisterSuccess';
import { FormType } from './dialogSlice';

export type DialogWidth = 'sm' | 'md' | 'lg';

export type ComponentProps = {
  handleChange?: ChangeEventHandler<HTMLInputElement>;
};

export type DialogDetail = {
  title: string;
  component: ({ handleChange }: ComponentProps) => ReactNode;
  width: DialogWidth;
};

type DialogMap = Map<FormType, DialogDetail>;

export const getDialogElement = (formType: FormType) => {
  const map: DialogMap = new Map([
    [
      FormType.register,
      {
        title: 'Sign Up',
        component: Register,
        width: 'md',
      },
    ],
    [
      FormType.registerSuccess,
      {
        title: 'Congratulation',
        component: RegisterSuccess,
        width: 'sm',
      },
    ],
    [
      FormType.login,
      {
        title: 'Log In',
        component: Login,
        width: 'sm',
      },
    ],
    [
      FormType.loginSuccess,
      {
        title: 'Congratulation',
        component: LoginSuccess,
        width: 'sm',
      },
    ],
    [
      FormType.orderDetail,
      {
        title: 'Order Details',
        component: OrderDetail,
        width: 'lg',
      },
    ],
  ]);

  return map.get(formType);
};
