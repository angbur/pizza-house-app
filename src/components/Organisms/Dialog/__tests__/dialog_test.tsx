import { renderWithProviders } from 'test/test-utils';
import Dialog from '../Dialog';
import { FormType } from '../dialogSlice';

describe('Dialog', () => {
  it('should renders dialog component correctly with register-component', () => {
    const initialState = {
      isOpen: true,
      formType: FormType.register,
    };

    const dialog = renderWithProviders(<Dialog />, {
      preloadedState: {
        dialog: initialState,
      },
    }).container;

    expect(dialog).toMatchSnapshot();
  });

  it('should renders dialog component correctly with login-component', () => {
    const initialState = {
      isOpen: true,
      formType: FormType.login,
    };

    const dialog = renderWithProviders(<Dialog />, {
      preloadedState: {
        dialog: initialState,
      },
    }).container;

    expect(dialog).toMatchSnapshot();
  });
});
