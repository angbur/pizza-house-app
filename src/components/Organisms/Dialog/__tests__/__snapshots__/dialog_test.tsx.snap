// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog should renders dialog component correctly with login-component 1`] = `
<div>
  <div
    class="sc-eDWCr vNJod"
  >
    <div
      class="sc-bqWxrE iDiwNo"
    />
    <div
      class="sc-ksBlkl hPuuqZ"
      style="max-width: 400px;"
    >
      <div
        class="dialog-header"
      >
        <button
          class="sc-bcXHqe sc-pyfCe feNVH hIOFrL icon-button btn-lg header_actions_button--mobile button"
          style="padding: 0px; font-size: 1.05rem;"
        >
          <img
            alt=""
            height="12px"
            src="cancel-cross.svg"
          />
        </button>
        <h2
          class="dialog-title"
          style="font-family: Source Sans Pro; font-weight: 600; font-size: 1.6rem; line-height: 1.6; letter-spacing: 1px; color: rgb(21, 21, 21); text-transform: uppercase;"
        >
          Log In
        </h2>
      </div>
      <div>
        <form>
          <div
            class="sc-kDvujY epQXUX"
          >
            <div
              class="sc-jrcTuL cwMVBS"
            >
              <label
                style="color: rgb(21, 21, 21);"
              >
                Login
              </label>
              <input
                name="login"
                placeholder="Your login"
                style="color: rgb(21, 21, 21);"
                type="text"
                value=""
              />
              <p
                class="paragraph"
                style="font-family: Source Sans Pro; font-weight: 400; font-size: 1rem; line-height: 1.6; letter-spacing: 0; color: rgb(231, 76, 60);"
              />
            </div>
            <div
              class="sc-jrcTuL cwMVBS"
            >
              <label
                style="color: rgb(21, 21, 21);"
              >
                Password
              </label>
              <input
                name="password"
                placeholder="Password"
                style="color: rgb(21, 21, 21);"
                type="password"
                value=""
              />
              <p
                class="paragraph"
                style="font-family: Source Sans Pro; font-weight: 400; font-size: 1rem; line-height: 1.6; letter-spacing: 0; color: rgb(231, 76, 60);"
              />
            </div>
          </div>
        </form>
      </div>
      <div
        class="sc-hBxehG fXFSFF"
      >
        <button
          class="sc-bcXHqe sc-gswNZR feNVH gHullA primary-light btn-md button"
          style="padding: 0.5850000000000001rem 1.125rem; font-size: 0.9500000000000001rem;"
        >
          Log In
        </button>
        <p
          class="paragraph"
          style="font-family: Source Sans Pro; font-weight: 400; font-size: 1rem; line-height: 1.6; letter-spacing: 0; color: rgb(21, 21, 21);"
        >
          If you do not have an account
          <button
            class="sc-bcXHqe sc-gKPRtg sc-iBYQkv feNVH hUzBlu YZIpJ button-text-light btn-sm button"
            style="padding: 0.52rem 1rem; font-size: 0.8500000000000001rem;"
          >
            SIGN UP
          </button>
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Dialog should renders dialog component correctly with register-component 1`] = `
<div>
  <div
    class="sc-eDWCr vNJod"
  >
    <div
      class="sc-bqWxrE iDiwNo"
    />
    <div
      class="sc-ksBlkl hPuuqZ"
      style="max-width: 440.00000000000006px;"
    >
      <div
        class="dialog-header"
      >
        <button
          class="sc-bcXHqe sc-pyfCe feNVH hIOFrL icon-button btn-lg header_actions_button--mobile button"
          style="padding: 0px; font-size: 1.05rem;"
        >
          <img
            alt=""
            height="12px"
            src="cancel-cross.svg"
          />
        </button>
        <h2
          class="dialog-title"
          style="font-family: Source Sans Pro; font-weight: 600; font-size: 1.6rem; line-height: 1.6; letter-spacing: 1px; color: rgb(21, 21, 21); text-transform: uppercase;"
        >
          Sign Up
        </h2>
      </div>
      <div>
        <form>
          <div
            class="sc-csuSiG iIwxMD"
          >
            <div
              class="sc-jrcTuL cwMVBS"
            >
              <label
                style="color: rgb(21, 21, 21);"
              >
                Login
              </label>
              <input
                name="login"
                placeholder="Your login"
                style="color: rgb(21, 21, 21);"
                type="text"
                value=""
              />
              <p
                class="paragraph"
                style="font-family: Source Sans Pro; font-weight: 400; font-size: 1rem; line-height: 1.6; letter-spacing: 0; color: rgb(231, 76, 60);"
              >
                Minimum 4 letters, at least one capital letter and one number
              </p>
            </div>
            <div
              class="sc-jrcTuL cwMVBS"
            >
              <label
                style="color: rgb(21, 21, 21);"
              >
                First name
              </label>
              <input
                name="firstName"
                placeholder="Your first name"
                style="color: rgb(21, 21, 21);"
                type="text"
                value=""
              />
              <p
                class="paragraph"
                style="font-family: Source Sans Pro; font-weight: 400; font-size: 1rem; line-height: 1.6; letter-spacing: 0; color: rgb(231, 76, 60);"
              >
                Minimum 3 letters
              </p>
            </div>
            <div
              class="sc-jrcTuL cwMVBS"
            >
              <label
                style="color: rgb(21, 21, 21);"
              >
                Last name
              </label>
              <input
                name="lastName"
                placeholder="Your last name"
                style="color: rgb(21, 21, 21);"
                type="text"
                value=""
              />
              <p
                class="paragraph"
                style="font-family: Source Sans Pro; font-weight: 400; font-size: 1rem; line-height: 1.6; letter-spacing: 0; color: rgb(231, 76, 60);"
              >
                Minimum 3 letters
              </p>
            </div>
            <div
              class="sc-jrcTuL cwMVBS"
            >
              <label
                style="color: rgb(21, 21, 21);"
              >
                Email
              </label>
              <input
                name="email"
                placeholder="Your email"
                style="color: rgb(21, 21, 21);"
                type="email"
                value=""
              />
              <p
                class="paragraph"
                style="font-family: Source Sans Pro; font-weight: 400; font-size: 1rem; line-height: 1.6; letter-spacing: 0; color: rgb(231, 76, 60);"
              />
            </div>
            <div
              class="sc-jrcTuL cwMVBS"
            >
              <label
                style="color: rgb(21, 21, 21);"
              >
                Password
              </label>
              <input
                name="password"
                placeholder="Password"
                style="color: rgb(21, 21, 21);"
                type="password"
                value=""
              />
              <p
                class="paragraph"
                style="font-family: Source Sans Pro; font-weight: 400; font-size: 1rem; line-height: 1.6; letter-spacing: 0; color: rgb(231, 76, 60);"
              >
                At least 8 characters, one capital letter and one number
              </p>
            </div>
          </div>
        </form>
      </div>
      <div
        class="sc-hBxehG fXFSFF"
      >
        <button
          class="sc-bcXHqe sc-gswNZR feNVH gHullA primary-light btn-md button"
          style="padding: 0.5850000000000001rem 1.125rem; font-size: 0.9500000000000001rem;"
        >
          Create Account
        </button>
        <p
          class="paragraph"
          style="font-family: Source Sans Pro; font-weight: 400; font-size: 1rem; line-height: 1.6; letter-spacing: 0; color: rgb(21, 21, 21);"
        >
          If you already have an account just
          <button
            class="sc-bcXHqe sc-gKPRtg sc-iBYQkv feNVH hUzBlu YZIpJ button-text-light btn-sm button"
            style="padding: 0.52rem 1rem; font-size: 0.8500000000000001rem;"
          >
            SIGN IN
          </button>
        </p>
      </div>
    </div>
  </div>
</div>
`;
